File: Hashtable/tasks/twoSum.java
/**
 * 两数之和
   Given nums = [2, 7, 11, 15], target = 9,

   因为 nums[0] + nums[1] = 2 + 7 = 9,
   返回 [0, 1].

   使用哈希表 hashMap 的查找是 O(1)，时间复杂度就是O(N)
 */

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int i = 0; i< nums.length; i++) {
            map.put(nums[i], i);
        }
        
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement) && map.get(complement) != i) {
                return new int[] {i, map.get(complement)}; 
            }
        }
        
        throw new IllegalArgumentException("No two sum solution");
    }
}

File: Hashtable/tasks/twoSum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum1 = function(nums, target) {
  // 暴力解法 不可取，时间复杂度 O(N^2)
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[j] == target - nums[i]) {
        return [i, j];
      }
    }
  }
};

/**
 * 使用ES6 Map实现两数之和
 */
var twoSum = function(nums, target) {
  map = new Map();

  for (let i = 0; i < nums.length; i++) {
    map.set(nums[i], i);
  }

  for (let i = 0; i < nums.length; i++) {
    // 在 Map 中查找 complement ，hashMap的查找效率很高，时间复杂度O(1)
    let complement = target - nums[i];
    if (map.has(complement) && map.get(complement) != i) {
      return [i, map.get(complement)];
    }
  }

  throw Error("No two sum solution");
};

console.log(twoSum([1, 2, 3], 4));
